File: 0589-n-ary-tree-preorder-traversal/0589-n-ary-tree-preorder-traversal.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> children;

    Node() {}

    Node(int _val) {
        val = _val;
    }

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};
*/

class Solution {
public:
    void preord(Node* root,vector<int>& ans){
        if(root!=NULL){
         ans.push_back(root->val);
            for(int i=0;i<root->children.size();i++){
                preord(root->children[i],ans);
            }
        }
    }
    vector<int> preorder(Node* root) {
        vector<int> ans;
        preord(root,ans);
        return ans;
}
};

File: 0589-n-ary-tree-preorder-traversal/0589-n-ary-tree-preorder-traversal.cs
public class Solution {
    public IList<int> Preorder(Node root) {
        IList<int> preOrder = new List<int>();
        Stack<Node> s = new Stack<Node>();
        Node tmp;

        s.Push(root);
        while (s.TryPop(out tmp) && tmp != null)
        {
            preOrder.Add(tmp.val);
            for (int i = tmp.children.Count - 1; i >= 0; i--)
            {
                s.Push(tmp.children[i]);
            }
        }
        return preOrder;
    }
}